File: src/main/java/brs/services/impl/DGSGoodsStoreServiceImpl.java
package brs.services.impl;

import brs.*;
import brs.DigitalGoodsStore.Event;
import brs.DigitalGoodsStore.Goods;
import brs.DigitalGoodsStore.Purchase;
import brs.crypto.EncryptedData;
import brs.db.BurstKey;
import brs.db.BurstKey.LongKeyFactory;
import brs.db.VersionedEntityTable;
import brs.db.VersionedValuesTable;
import brs.db.store.DigitalGoodsStoreStore;
import brs.services.AccountService;
import brs.services.DGSGoodsStoreService;
import brs.util.Convert;
import brs.util.Listeners;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.function.Consumer;

public class DGSGoodsStoreServiceImpl implements DGSGoodsStoreService {

  private final DependencyProvider dp;
  private final VersionedValuesTable<Purchase, EncryptedData> feedbackTable;
  private final VersionedValuesTable<Purchase, String> publicFeedbackTable;

  private final VersionedEntityTable<Goods> goodsTable;
  private final VersionedEntityTable<Purchase> purchaseTable;
  private final LongKeyFactory<Goods> goodsDbKeyFactory;
  private final LongKeyFactory<Purchase> purchaseDbKeyFactory;

  private final Listeners<Goods,Event> goodsListeners = new Listeners<>();

  private final Listeners<Purchase,Event> purchaseListeners = new Listeners<>();

  public DGSGoodsStoreServiceImpl(DependencyProvider dp) {
    this.dp = dp;
    this.goodsTable = dp.digitalGoodsStoreStore.getGoodsTable();
    this.purchaseTable = dp.digitalGoodsStoreStore.getPurchaseTable();
    this.goodsDbKeyFactory = dp.digitalGoodsStoreStore.getGoodsDbKeyFactory();
    this.purchaseDbKeyFactory = dp.digitalGoodsStoreStore.getPurchaseDbKeyFactory();
    this.feedbackTable = dp.digitalGoodsStoreStore.getFeedbackTable();
    this.publicFeedbackTable = dp.digitalGoodsStoreStore.getPublicFeedbackTable();
  }

  @Override
  public boolean addGoodsListener(Consumer<Goods> listener, Event eventType) {
    return goodsListeners.addListener(listener, eventType);
  }

  @Override
  public boolean removeGoodsListener(Consumer<Goods> listener, Event eventType) {
    return goodsListeners.removeListener(listener, eventType);
  }

  @Override
  public boolean addPurchaseListener(Consumer<Purchase> listener, Event eventType) {
    return purchaseListeners.addListener(listener, eventType);
  }

  @Override
  public boolean removePurchaseListener(Consumer<Purchase> listener, Event eventType) {
    return purchaseListeners.removeListener(listener, eventType);
  }

  @Override
  public Goods getGoods(long goodsId) {
    return goodsTable.get(goodsDbKeyFactory.newKey(goodsId));
  }

  @Override
  public Collection<Goods> getAllGoods(int from, int to) {
    return goodsTable.getAll(from, to);
  }

  @Override
  public Collection<Goods> getGoodsInStock(int from, int to) {
    return dp.digitalGoodsStoreStore.getGoodsInStock(from, to);
  }

  @Override
  public Collection<Goods> getSellerGoods(final long sellerId, final boolean inStockOnly, int from, int to) {
    return dp.digitalGoodsStoreStore.getSellerGoods(sellerId, inStockOnly, from, to);
  }

  @Override
  public Collection<Purchase> getAllPurchases(int from, int to) {
    return purchaseTable.getAll(from, to);
  }

  @Override
  public Collection<Purchase> getSellerPurchases(long sellerId, int from, int to) {
    return dp.digitalGoodsStoreStore.getSellerPurchases(sellerId, from, to);
  }

  @Override
  public Collection<Purchase> getBuyerPurchases(long buyerId, int from, int to) {
    return dp.digitalGoodsStoreStore.getBuyerPurchases(buyerId, from, to);
  }

  @Override
  public Collection<Purchase> getSellerBuyerPurchases(final long sellerId, final long buyerId, int from, int to) {
    return dp.digitalGoodsStoreStore.getSellerBuyerPurchases(sellerId, buyerId, from, to);
  }

  @Override
  public Collection<Purchase> getPendingSellerPurchases(final long sellerId, int from, int to) {
    return dp.digitalGoodsStoreStore.getPendingSellerPurchases(sellerId, from, to);
  }

  @Override
  public Purchase getPurchase(long purchaseId) {
    return purchaseTable.get(purchaseDbKeyFactory.newKey(purchaseId));
  }

  @Override
  public void changeQuantity(long goodsId, int deltaQuantity, boolean allowDelisted) {
    Goods goods = goodsTable.get(goodsDbKeyFactory.newKey(goodsId));
    if (allowDelisted || ! goods.isDelisted()) {
      goods.changeQuantity(deltaQuantity);
      goodsTable.insert(goods);
      goodsListeners.accept(goods, Event.GOODS_QUANTITY_CHANGE);
    } else {
      throw new IllegalStateException("Can't change quantity of delisted goods");
    }
  }

  @Override
  public void purchase(Transaction transaction, Attachment.DigitalGoodsPurchase attachment) {
    Goods goods = goodsTable.get(goodsDbKeyFactory.newKey(attachment.getGoodsId()));
    if (! goods.isDelisted() && attachment.getQuantity() <= goods.getQuantity() && attachment.getPriceNQT() == goods.getPriceNQT()
        && attachment.getDeliveryDeadlineTimestamp() > dp.blockchain.getLastBlock().getTimestamp()) {
      changeQuantity(goods.getId(), -attachment.getQuantity(), false);
      addPurchase(transaction, attachment, goods.getSellerId());
    } else {
      Account buyer = dp.accountService.getAccount(transaction.getSenderId());
      dp.accountService.addToUnconfirmedBalanceNQT(buyer, Convert.INSTANCE.safeMultiply(attachment.getQuantity(), attachment.getPriceNQT()));
      // restoring the unconfirmed balance if purchase not successful, however buyer still lost the transaction fees
    }
  }

  @Override
  public void addPurchase(Transaction transaction, Attachment.DigitalGoodsPurchase attachment, long sellerId) {
    Purchase purchase = new Purchase(dp, transaction, attachment, sellerId);
    purchaseTable.insert(purchase);
    purchaseListeners.accept(purchase, Event.PURCHASE);
  }

  @Override
  public void listGoods(Transaction transaction, Attachment.DigitalGoodsListing attachment) {
    BurstKey dbKey = goodsDbKeyFactory.newKey(transaction.getId());
    Goods goods = new Goods(dbKey, transaction, attachment);
    goodsTable.insert(goods);
    goodsListeners.accept(goods, Event.GOODS_LISTED);
  }

  @Override
  public void delistGoods(long goodsId) {
    Goods goods = goodsTable.get(goodsDbKeyFactory.newKey(goodsId));
    if (! goods.isDelisted()) {
      goods.setDelisted(true);
      goodsTable.insert(goods);
      goodsListeners.accept(goods, Event.GOODS_DELISTED);
    } else {
      throw new IllegalStateException("Goods already delisted");
    }
  }

  @Override
  public void feedback(long purchaseId, Appendix.EncryptedMessage encryptedMessage, Appendix.Message message) {
    Purchase purchase = purchaseTable.get(purchaseDbKeyFactory.newKey(purchaseId));
    if (encryptedMessage != null) {
      purchase.addFeedbackNote(encryptedMessage.getEncryptedData());
      purchaseTable.insert(purchase);
      feedbackTable.insert(purchase, purchase.getFeedbackNotes());
    }
    if (message != null) {
      addPublicFeedback(purchase, Convert.INSTANCE.toString(message.getMessageBytes()));
    }
    purchaseListeners.accept(purchase, Event.FEEDBACK);
  }

  private void addPublicFeedback(Purchase purchase, String publicFeedback) {
    List<String> publicFeedbacks = purchase.getPublicFeedbacks();
    if (publicFeedbacks == null) {
      publicFeedbacks = new ArrayList<>();
    }
    publicFeedbacks.add(publicFeedback);
    purchase.setHasPublicFeedbacks(true);
    purchaseTable.insert(purchase);
    publicFeedbackTable.insert(purchase, publicFeedbacks);
  }

  @Override
  public void refund(long sellerId, long purchaseId, long refundNQT, Appendix.EncryptedMessage encryptedMessage) {
    Purchase purchase = purchaseTable.get(purchaseDbKeyFactory.newKey(purchaseId));
    Account seller = dp.accountService.getAccount(sellerId);
    dp.accountService.addToBalanceNQT(seller, -refundNQT);
    Account buyer = dp.accountService.getAccount(purchase.getBuyerId());
    dp.accountService.addToBalanceAndUnconfirmedBalanceNQT(buyer, refundNQT);
    if (encryptedMessage != null) {
      purchase.setRefundNote(encryptedMessage.getEncryptedData());
      purchaseTable.insert(purchase);
    }
    purchase.setRefundNQT(refundNQT);
    purchaseTable.insert(purchase);
    purchaseListeners.accept(purchase, Event.REFUND);
  }

  @Override
  public Collection<Purchase> getExpiredPendingPurchases(final int timestamp) {
    return dp.digitalGoodsStoreStore.getExpiredPendingPurchases(timestamp);
  }

  @Override
  public void changePrice(long goodsId, long priceNQT) {
    Goods goods = goodsTable.get(goodsDbKeyFactory.newKey(goodsId));
    if (! goods.isDelisted()) {
      goods.changePrice(priceNQT);
      goodsTable.insert(goods);
      goodsListeners.accept(goods, Event.GOODS_PRICE_CHANGE);
    } else {
      throw new IllegalStateException("Can't change price of delisted goods");
    }
  }

  @Override
  public void deliver(Transaction transaction, Attachment.DigitalGoodsDelivery attachment) {
    Purchase purchase = getPendingPurchase(attachment.getPurchaseId());
    if ( purchase == null ) {
      throw new RuntimeException("cant find purchase with id " + attachment.getPurchaseId());
    }
    setPending(purchase, false);
    long totalWithoutDiscount = Convert.INSTANCE.safeMultiply(purchase.getQuantity(), purchase.getPriceNQT());
    Account buyer = dp.accountService.getAccount(purchase.getBuyerId());
    dp.accountService.addToBalanceNQT(buyer, Convert.INSTANCE.safeSubtract(attachment.getDiscountNQT(), totalWithoutDiscount));
    dp.accountService.addToUnconfirmedBalanceNQT(buyer, attachment.getDiscountNQT());
    Account seller = dp.accountService.getAccount(transaction.getSenderId());
    dp.accountService.addToBalanceAndUnconfirmedBalanceNQT(seller, Convert.INSTANCE.safeSubtract(totalWithoutDiscount, attachment.getDiscountNQT()));
    purchase.setEncryptedGoods(attachment.getGoods(), attachment.goodsIsText());
    purchaseTable.insert(purchase);
    purchase.setDiscountNQT(attachment.getDiscountNQT());
    purchaseTable.insert(purchase);
    purchaseListeners.accept(purchase, Event.DELIVERY);
  }

  @Override
  public Purchase getPendingPurchase(long purchaseId) {
    Purchase purchase = getPurchase(purchaseId);
    return purchase == null || ! purchase.isPending() ? null : purchase;
  }

  @Override
  public void setPending(Purchase purchase, boolean pendingValue) {
    purchase.setPending(pendingValue);
    purchaseTable.insert(purchase);
  }

  public static class ExpiredPurchaseListener implements Consumer<Block> {
    private final DependencyProvider dp;

    public ExpiredPurchaseListener(DependencyProvider dp) {
      this.dp = dp;
    }

    @Override
    public void accept(Block block) {
      for (Purchase purchase : dp.digitalGoodsStoreService.getExpiredPendingPurchases(block.getTimestamp())) {
        Account buyer = dp.accountService.getAccount(purchase.getBuyerId());
        dp.accountService.addToUnconfirmedBalanceNQT(buyer, Convert.INSTANCE.safeMultiply(purchase.getQuantity(), purchase.getPriceNQT()));
        dp.digitalGoodsStoreService.changeQuantity(purchase.getGoodsId(), purchase.getQuantity(), true);
        dp.digitalGoodsStoreService.setPending(purchase, false);
      }
    }
  }
}

File: src/main/java/brs/services/impl/DGSGoodsStoreServiceImpl.kt
package brs.services.impl

import brs.*
import brs.DigitalGoodsStore.Event
import brs.DigitalGoodsStore.Goods
import brs.DigitalGoodsStore.Purchase
import brs.services.DGSGoodsStoreService
import brs.util.Listeners
import brs.util.convert.safeMultiply
import brs.util.convert.safeSubtract
import brs.util.convert.toUtf8String

class DGSGoodsStoreServiceImpl(private val dp: DependencyProvider) : DGSGoodsStoreService {
    private val feedbackTable = dp.digitalGoodsStoreStore.feedbackTable
    private val publicFeedbackTable = dp.digitalGoodsStoreStore.publicFeedbackTable

    private val goodsTable = dp.digitalGoodsStoreStore.goodsTable
    private val purchaseTable = dp.digitalGoodsStoreStore.purchaseTable
    private val goodsDbKeyFactory = dp.digitalGoodsStoreStore.goodsDbKeyFactory
    private val purchaseDbKeyFactory = dp.digitalGoodsStoreStore.purchaseDbKeyFactory

    private val goodsListeners = Listeners<Goods, Event>()

    private val purchaseListeners = Listeners<Purchase, Event>()

    override suspend fun addGoodsListener(listener: suspend (Goods) -> Unit, eventType: Event) {
        goodsListeners.addListener(eventType, listener)
    }

    override suspend fun addPurchaseListener(listener: suspend (Purchase) -> Unit, eventType: Event) {
        purchaseListeners.addListener(eventType, listener)
    }

    override suspend fun getGoods(goodsId: Long): Goods? {
        return goodsTable.get(goodsDbKeyFactory.newKey(goodsId))
    }

    override suspend fun getAllGoods(from: Int, to: Int): Collection<Goods> {
        return goodsTable.getAll(from, to)
    }

    override suspend fun getGoodsInStock(from: Int, to: Int): Collection<Goods> {
        return dp.digitalGoodsStoreStore.getGoodsInStock(from, to)
    }

    override suspend fun getSellerGoods(sellerId: Long, inStockOnly: Boolean, from: Int, to: Int): Collection<Goods> {
        return dp.digitalGoodsStoreStore.getSellerGoods(sellerId, inStockOnly, from, to)
    }

    override suspend fun getAllPurchases(from: Int, to: Int): Collection<Purchase> {
        return purchaseTable.getAll(from, to)
    }

    override suspend fun getSellerPurchases(sellerId: Long, from: Int, to: Int): Collection<Purchase> {
        return dp.digitalGoodsStoreStore.getSellerPurchases(sellerId, from, to)
    }

    override suspend fun getBuyerPurchases(buyerId: Long, from: Int, to: Int): Collection<Purchase> {
        return dp.digitalGoodsStoreStore.getBuyerPurchases(buyerId, from, to)
    }

    override suspend fun getSellerBuyerPurchases(sellerId: Long, buyerId: Long, from: Int, to: Int): Collection<Purchase> {
        return dp.digitalGoodsStoreStore.getSellerBuyerPurchases(sellerId, buyerId, from, to)
    }

    override suspend fun getPendingSellerPurchases(sellerId: Long, from: Int, to: Int): Collection<Purchase> {
        return dp.digitalGoodsStoreStore.getPendingSellerPurchases(sellerId, from, to)
    }

    override suspend fun getPurchase(purchaseId: Long): Purchase? {
        return purchaseTable.get(purchaseDbKeyFactory.newKey(purchaseId))
    }

    override suspend fun changeQuantity(goodsId: Long, deltaQuantity: Int, allowDelisted: Boolean) {
        val goods = goodsTable.get(goodsDbKeyFactory.newKey(goodsId))!!
        if (allowDelisted || !goods.isDelisted) {
            goods.changeQuantity(deltaQuantity)
            goodsTable.insert(goods)
            goodsListeners.accept(Event.GOODS_QUANTITY_CHANGE, goods)
        } else {
            throw IllegalStateException("Can't change quantity of delisted goods")
        }
    }

    override suspend fun purchase(transaction: Transaction, attachment: Attachment.DigitalGoodsPurchase) {
        val goods = goodsTable.get(goodsDbKeyFactory.newKey(attachment.goodsId))!!
        if (!goods.isDelisted && attachment.quantity <= goods.quantity && attachment.priceNQT == goods.priceNQT
                && attachment.deliveryDeadlineTimestamp > dp.blockchain.lastBlock.timestamp) {
            changeQuantity(goods.id, -attachment.quantity, false)
            addPurchase(transaction, attachment, goods.sellerId)
        } else {
            val buyer = dp.accountService.getAccount(transaction.senderId)!!
            dp.accountService.addToUnconfirmedBalanceNQT(buyer, attachment.quantity.toLong().safeMultiply(attachment.priceNQT))
            // restoring the unconfirmed balance if purchase not successful, however buyer still lost the transaction fees
        }
    }

    override suspend fun addPurchase(transaction: Transaction, attachment: Attachment.DigitalGoodsPurchase, sellerId: Long) {
        val purchase = Purchase(dp, transaction, attachment, sellerId)
        purchaseTable.insert(purchase)
        purchaseListeners.accept(Event.PURCHASE, purchase)
    }

    override suspend fun listGoods(transaction: Transaction, attachment: Attachment.DigitalGoodsListing) {
        val dbKey = goodsDbKeyFactory.newKey(transaction.id)
        val goods = Goods(dbKey, transaction, attachment)
        goodsTable.insert(goods)
        goodsListeners.accept(Event.GOODS_LISTED, goods)
    }

    override suspend fun delistGoods(goodsId: Long) {
        val goods = goodsTable.get(goodsDbKeyFactory.newKey(goodsId))!!
        if (!goods.isDelisted) {
            goods.isDelisted = true
            goodsTable.insert(goods)
            goodsListeners.accept(Event.GOODS_DELISTED, goods)
        } else {
            throw IllegalStateException("Goods already delisted")
        }
    }

    override suspend fun feedback(purchaseId: Long, encryptedMessage: Appendix.EncryptedMessage?, message: Appendix.Message?) {
        val purchase = purchaseTable.get(purchaseDbKeyFactory.newKey(purchaseId))!!
        if (encryptedMessage != null) {
            purchase.addFeedbackNote(encryptedMessage.encryptedData)
            purchaseTable.insert(purchase)
            feedbackTable.insert(purchase, purchase.feedbackNotes!!)
        }
        if (message != null) {
            addPublicFeedback(purchase, message.messageBytes.toUtf8String())
        }
        purchaseListeners.accept(Event.FEEDBACK, purchase)
    }

    private suspend fun addPublicFeedback(purchase: Purchase, publicFeedback: String) {
        var publicFeedbacks: MutableList<String>? = purchase.publicFeedbacks
        if (publicFeedbacks == null) {
            publicFeedbacks = mutableListOf()
        }
        publicFeedbacks.add(publicFeedback)
        purchase.setHasPublicFeedbacks(true)
        purchaseTable.insert(purchase)
        publicFeedbackTable.insert(purchase, publicFeedbacks)
    }

    override suspend fun refund(sellerId: Long, purchaseId: Long, refundNQT: Long, encryptedMessage: Appendix.EncryptedMessage?) {
        val purchase = purchaseTable.get(purchaseDbKeyFactory.newKey(purchaseId))!!
        val seller = dp.accountService.getAccount(sellerId)!!
        dp.accountService.addToBalanceNQT(seller, -refundNQT)
        val buyer = dp.accountService.getAccount(purchase.buyerId)!!
        dp.accountService.addToBalanceAndUnconfirmedBalanceNQT(buyer, refundNQT)
        if (encryptedMessage != null) {
            purchase.refundNote = encryptedMessage.encryptedData
            purchaseTable.insert(purchase)
        }
        purchase.refundNQT = refundNQT
        purchaseTable.insert(purchase)
        purchaseListeners.accept(Event.REFUND, purchase)
    }

    override suspend fun getExpiredPendingPurchases(timestamp: Int): Collection<Purchase> {
        return dp.digitalGoodsStoreStore.getExpiredPendingPurchases(timestamp)
    }

    override suspend fun changePrice(goodsId: Long, priceNQT: Long) {
        val goods = goodsTable.get(goodsDbKeyFactory.newKey(goodsId))!!
        if (!goods.isDelisted) {
            goods.changePrice(priceNQT)
            goodsTable.insert(goods)
            goodsListeners.accept(Event.GOODS_PRICE_CHANGE, goods)
        } else {
            throw IllegalStateException("Can't change price of delisted goods")
        }
    }

    override suspend fun deliver(transaction: Transaction, attachment: Attachment.DigitalGoodsDelivery) {
        val purchase = getPendingPurchase(attachment.purchaseId)
                ?: throw RuntimeException("cant find purchase with id " + attachment.purchaseId)
        setPending(purchase, false)
        val totalWithoutDiscount = purchase.quantity.toLong().safeMultiply(purchase.priceNQT)
        val buyer = dp.accountService.getAccount(purchase.buyerId)!!
        dp.accountService.addToBalanceNQT(buyer, attachment.discountNQT.safeSubtract(totalWithoutDiscount))
        dp.accountService.addToUnconfirmedBalanceNQT(buyer, attachment.discountNQT)
        val seller = dp.accountService.getAccount(transaction.senderId)!!
        dp.accountService.addToBalanceAndUnconfirmedBalanceNQT(seller, totalWithoutDiscount.safeSubtract(attachment.discountNQT))
        purchase.setEncryptedGoods(attachment.goods, attachment.goodsIsText())
        purchaseTable.insert(purchase)
        purchase.discountNQT = attachment.discountNQT
        purchaseTable.insert(purchase)
        purchaseListeners.accept(Event.DELIVERY, purchase)
    }

    override suspend fun getPendingPurchase(purchaseId: Long): Purchase? {
        val purchase = getPurchase(purchaseId)
        return if (purchase == null || !purchase.isPending) null else purchase
    }

    override suspend fun setPending(purchase: Purchase, pendingValue: Boolean) {
        purchase.isPending = pendingValue
        purchaseTable.insert(purchase)
    }

    companion object {
        fun expiredPurchaseListener(dp: DependencyProvider): suspend (Block) -> Unit = { block ->
            for (purchase in dp.digitalGoodsStoreService.getExpiredPendingPurchases(block.timestamp)) {
                val buyer = dp.accountService.getAccount(purchase.buyerId)!!
                dp.accountService.addToUnconfirmedBalanceNQT(
                    buyer,
                    purchase.quantity.toLong().safeMultiply(purchase.priceNQT)
                )
                dp.digitalGoodsStoreService.changeQuantity(purchase.goodsId, purchase.quantity, true)
                dp.digitalGoodsStoreService.setPending(purchase, false)
            }
        }
    }
}
